kmath-core/src/commonMain/kotlin/scientifik/kmath/operations/NumberAlgebra.kt

```diff
@@ -54,93 +63,109 @@ inline class Real(val value: Double) : FieldElement<Double, Real, RealField> {
-    override inline fun atan(arg: Float) = kotlin.math.atan(arg)
+    override val one: Float
+        get() = 1.0f
 
-    override inline fun power(arg: Float, pow: Number) = arg.pow(pow.toFloat())
+    override inline fun add(a: Float, b: Float): Float = a + b
+    override inline fun multiply(a: Float, k: Number): Float = a * k.toFloat()
 
-    override inline fun exp(arg: Float) = kotlin.math.exp(arg)
-    override inline fun ln(arg: Float) = kotlin.math.ln(arg)
+    override inline fun multiply(a: Float, b: Float): Float = a * b
 
-    override inline fun norm(arg: Float) = abs(arg)
+    override inline fun divide(a: Float, b: Float): Float = a / b
 
-    override inline fun Float.unaryMinus() = -this
+    override inline fun sin(arg: Float): Float = kotlin.math.sin(arg)
+    override inline fun cos(arg: Float): Float = kotlin.math.cos(arg)
+    override inline fun tan(arg: Float): Float = kotlin.math.tan(arg)
+    override inline fun acos(arg: Float): Float = kotlin.math.acos(arg)
+    override inline fun asin(arg: Float): Float = kotlin.math.asin(arg)
+    override inline fun atan(arg: Float): Float = kotlin.math.atan(arg)
 
-    override inline fun Float.plus(b: Float) = this + b
+    override inline fun sinh(arg: Float): Float = kotlin.math.sinh(arg)
+    override inline fun cosh(arg: Float): Float = kotlin.math.cosh(arg)
+    override inline fun tanh(arg: Float): Float = kotlin.math.tanh(arg)
+    override inline fun asinh(arg: Float): Float = kotlin.math.asinh(arg)
+    override inline fun acosh(arg: Float): Float = kotlin.math.acosh(arg)
+    override inline fun atanh(arg: Float): Float = kotlin.math.atanh(arg)
 
-    override inline fun Float.minus(b: Float) = this - b
+    override inline fun power(arg: Float, pow: Number): Float = arg.kpow(pow.toFloat())
+    override inline fun exp(arg: Float): Float = kotlin.math.exp(arg)
+    override inline fun ln(arg: Float): Float = kotlin.math.ln(arg)
 
-    override inline fun Float.times(b: Float) = this * b
+    override inline fun norm(arg: Float): Float = abs(arg)
 
-    override inline fun Float.div(b: Float) = this / b
+    override inline fun Float.unaryMinus(): Float = -this
+    override inline fun Float.plus(b: Float): Float = this + b
+    override inline fun Float.minus(b: Float): Float = this - b
+    override inline fun Float.times(b: Float): Float = this * b
+    override inline fun Float.div(b: Float): Float = this / b
 }
 
 /**
  * A field for [Int] without boxing. Does not produce corresponding field element
  */
 @Suppress("EXTENSION_SHADOWED_BY_MEMBER", "OVERRIDE_BY_INLINE", "NOTHING_TO_INLINE")
 object IntRing : Ring<Int>, Norm<Int, Int> {
-    override val zero: Int = 0
-    override inline fun add(a: Int, b: Int) = a + b
-    override inline fun multiply(a: Int, b: Int) = a * b
-    override inline fun multiply(a: Int, k: Number) = k.toInt() * a
-    override val one: Int = 1
+    override val zero: Int
```

Comment:
Does not make sense to add virtual call here. There is not difference, I guess.

Reply:
The previous property in Java was like:
```private int zero = 0; public int getZero() { return zero; }```
This one doesn't store zero and doesn't spend any resources to instantiate the value each time because `0` is primitive constant: `public int getZero() { return 0; }`

Reply:
So, this implementation is 4 bytes more efficient than previous :)

Reply:
It is a singleton, another 4 bytes do not matter. You have an integer allocation on each call though.

Reply:
No

Reply:
`getfield...` and `iconst_0` produce the equal allocation to the operand stack

Reply:
OK, you are correct in case of primitives.